distribution/docs/src/main/jdocs/content/_developingComponents/developing-plugins-contents.adoc

```diff
@@ -0,0 +1,403 @@
++
+. Import the ${branding} interface packages to the bundle manifest (in addition to any other required packages). +
+`Import-Package: ddf.catalog,ddf.catalog.plugin`
++
+. Export the service to the OSGi registry. +
+*Blueprint descriptor example*
+`<service ref="[[SamplePreIngestPlugin ]]"interface="ddf.catalog.plugin.PreIngestPlugin" />`
+
+====== Pre-Ingest Plugin Failure Behavior
+
+In the event that this Catalog Plugin cannot operate but does not wish to fail the transaction, a `PluginExecutionException` will be thrown.
+For any other Exceptions, the Catalog will "fail safe" and the Operation will be cancelled.
+If processing is to be explicitly stopped, a `StopProcessingException` will be thrown.
+
+====== Invoking Pre-Ingest Plugins
+
+Pre-Ingest plugins are invoked serially, prioritized by descending OSGi service ranking. 
+The plugin with the highest service ranking will be executed first. 
+
+The output of a Pre-Ingest plugin is sent to the next Pre-Ingest plugin, until all have executed and the ingest operation is sent to the requested Source.
+
+===== Implementing Post-Ingest Plugins
+
+Develop a custom Post-Ingest Plugin.
+
+. Create a Java class that implements `PostIngestPlugin`. +
+`public class SamplePostIngestPlugin implements ddf.catalog.plugin.PostIngestPlugin`
+
+. Implement the required methods. +
+ * `public CreateResponse process(CreateResponse input) *throws* PluginExecutionException;`
+ * `public UpdateResponse process(UpdateResponse input) *throws* PluginExecutionException;`
+ * `public DeleteResponse process(DeleteResponse input) *throws* PluginExecutionException;`
+
+. Import the ${branding} interface packages to the bundle manifest (in addition to any other required packages). +
+`Import-Package: ddf.catalog,ddf.catalog.plugin`
+
+. Export the service to the OSGi registry. +
+*Blueprint descriptor example*
+`<service ref="[[SamplePostIngestPlugin ]]"interface="ddf.catalog.plugin.PostIngestPlugin" />`
+
+====== Post-Ingest Plugin Failure Behavior
+
+In the event that this Catalog Plugin cannot operate but does not wish to fail the transaction, a `PluginExecutionException` will be thrown.
+
+====== Invoking Post-Ingest Plugins
+
+Post-Ingest plugins are invoked serially in rank order.
+
+===== Implementing Pre-Query Plugins
+
+Develop a custom Pre-Query Plugin
+
+. Create a Java class that implements `PreQueryPlugin`. +
+`public class SamplePreQueryPlugin implements ddf.catalog.plugin.PreQueryPlugin`
+
+. Implement the required method. +
+`public QueryRequest process(QueryRequest input) *throws* PluginExecutionException, StopProcessingException;`
+. Import the ${branding} interface packages to the bundle manifest (in addition to any other required packages). +
+`Import-Package: ddf.catalog,ddf.catalog.plugin`
+
+. Export the service to the OSGi registry. +
+`<service ref="[[SamplePreQueryPlugin]]"interface="ddf.catalog.plugin.PreQueryPlugin" />`
+
+=====  Pre-Query Plugins Failure Behavior
```

Comment:
❓ Pretty sure all the plugins have the same failure behavior. Could we just explain how they all work at the beginning instead of repeating this for each one?